### PR TITLE
Handle status monitor container reload

### PR DESCRIPTION
**Pull request by dteplits**
Flow description:

1. When  status monitor container is loaded, it creates a zone file and initializes it with the header and A records. In this case SOA serial initial value will be 1
2. Each time status monitor updates zone file, it increments SOA serial value
3. Auto plugin inside Corefile is responsible to reload zone file newest data into the DNS when it identifies that SOA serial value was changed

PR adds the below logic:

If status monitor container was reloaded, on its startup it should check whether zone file already exist.
If yes, it should continue the existing SOA serial sequence instead of setting it back to 1. 
The reason for that is the below edge case:

- If zone file already exist and holds SOA serial = 1 (for example) and status monitor container was restarted
- Then even after status monitor will override zone file content with the new data, the SOA serial will remain the same (=1)
- Thus it will not be reloaded by the auto plugin since the SOA serial didn't change

**Release note:**

```release-note
NONE
```

### CODE REVIEW

File: pkg/zonemgr/zone_file_cache.go [hunk 1]
```diff
@@ -55,10 +56,33 @@ func (zoneFileCache *ZoneFileCache) prepare() {
 	zoneFileCache.generateHeaderSuffix()
 	zoneFileCache.soaSerial = 0
```
Comment (oshoval) [inline pkg/zonemgr/zone_file_cache.go:57]: please read the file here, and set the serial either 0 or whatever the file has if exists
then no need this part
https://github.com/kubevirt/kubesecondarydns/pull/45/files#diff-91aeb36dc70c24eda218702f6964e34e7cbb3c7f29cce4f45ed3b09b6f59cb11R44

File: pkg/zonemgr/zone_file_cache.go [hunk 2]
```diff
@@ -55,10 +56,33 @@ func (zoneFileCache *ZoneFileCache) prepare() {
...
+		return 0
+	}
+	soaSerial := strings.TrimSpace(contentStr[ind1+1 : ind1+ind2])
+	if soaSerialInt, err := strconv.Atoi(soaSerial); err != nil {
+		return 0
+	} else {
+		return soaSerialInt
+	}
+}
+
```
Comment (oshoval) [inline pkg/zonemgr/zone_file_cache.go:85]: there might be a way using regex to fetch it, will be nice if we can check please

**Review by oshoval — COMMENTED** [review]
Thanks
partial review

Please consider updating PR desc with more precise flow that cause the dead end

File: pkg/zonemgr/zone_file_cache.go (hunk 1 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file_cache.go:57]: It is fine that there will be an helper function that reads the file in order to determine the SOA to start from.
It is better imo than to create the header twice

File: pkg/zonemgr/zone_manager.go [hunk 3]
```diff
@@ -40,6 +40,10 @@ func (zoneMgr *ZoneManager) prepare() {
 	zoneMgr.zoneFileCache = NewZoneFileCache(nameServerIP, domain)
 
 	zoneMgr.zoneFile = NewZoneFile(zoneFileName + zoneMgr.zoneFileCache.domain)
+
+	if content, _ := zoneMgr.zoneFile.readFile(); content != nil && len(content) > 0 {
```
Comment (oshoval) [inline pkg/zonemgr/zone_manager.go:44]: please verify that error is not nil as well
if it is !=nil it is allowed to only be "not found"

File: pkg/zonemgr/zone_file_cache.go (hunk 2 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file_cache.go:85]: Btw according https://www.ietf.org/rfc/rfc1035.txt you can add comments (start with `;`)
if the first line will be a comment with the serial, you can read just that line instead the whole file and get the serial

File: pkg/zonemgr/zone_file_cache.go (hunk 1 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go]: done

File: pkg/zonemgr/zone_file_cache.go (hunk 2 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go]: Changed to using regex.
Regarding the "reading line instead of the whole file", I can add it in a different PR

File: pkg/zonemgr/zone_file_cache_test.go [hunk 4]
```diff
@@ -38,6 +41,20 @@ var _ = Describe("cached zone file content maintenance", func() {
...
+		When("zone file already exist", func() {
+			BeforeEach(func() {
+				Expect(os.Mkdir("zones", 0777)).To(Succeed())
+				Expect(os.WriteFile(zoneFileName, []byte(content), 0777)).To(Succeed())
+			})
+			AfterEach(func() {
+				Expect(os.RemoveAll("zones")).To(Succeed())
+			})
+
+			It("should set SOA serial equal to existing value", func() {
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_file_cache_test.go:54]: I think the method the has to be tested is the prive one `NewZoneFileCache` and not an internal method.
I would call `NewZoneFileCache`and check the `header`. Same as the other tests.

**Review by AlonaKaplan — COMMENTED** [review]
General comment that I had to give on previous PRs (and can be fixed on a follow-up PR). I would change the code so there will be a clear API between ZoneFile, ZoneFileCache and Zonemanager.
I mean, treat them as if they are in separate packages and there is no access to prived methods and data members.
I believe it will make the code much more readable and safe.

File: pkg/zonemgr/zone_file_cache.go [hunk 5]
```diff
@@ -53,12 +57,32 @@ func (zoneFileCache *ZoneFileCache) prepare() {
...
+		if soaSerialInt, err := strconv.Atoi(soaSerial); err == nil {
+			return soaSerialInt
+		} else {
+			return 0
+		}
+	} else {
+		return 0
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go:85]: I'm doing pretty match the same:
your example:
re := regexp.MustCompile(`/user/([0-9]+)/edit$`)
x := re.FindStringSubmatch(s)
my code:
soaSerialReg = regexp.MustCompile("\\([0-9]+ ")
result := soaSerialReg.Find(content)

In this case regexp.Find does the job, I'm not sure why we need to use FindStringSubmatch() function.
The rest of the code removes unnecessary white spaces (if any) and converts string into a numeric.
So I'm not sure what exactly can/should be improved here.

[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/zonemgr/zone_file_cache.go:85]: Its instead TrimSpace and Atoi
you can give a regex, and it will do everything else for you, giving you the chunk by its own

File: pkg/zonemgr/zone_file.go [hunk 6]
```diff
@@ -19,3 +17,7 @@ func NewZoneFile(fileName string) *ZoneFile {
 func (zoneFile *ZoneFile) writeFile(content string) (err error) {
 	return os.WriteFile(zoneFile.zoneFileFullName, []byte(content), zoneFilePerm)
 }
+
+func readFile(zoneFileName string) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:21]: Because it's a "util" function,  I use it inside ZoneFileCache in order to fetch SoaSerial. 
In the previous version this method was part of ZoneFile and ZoneManager did the job since it holds ZoneFile instance. 
But Or suggested to move this logic into ZoneFileCache file and use readFile as a helper function.
Here is the conversation:
https://github.com/kubevirt/kubesecondarydns/pull/45#discussion_r1043114185

I can create a util file for readFile function, I just wanted to spare creating another file just for one method that anyway matches ZoneFile functionality

File: pkg/zonemgr/zone_file_cache.go [hunk 7]
```diff
@@ -53,12 +57,32 @@ func (zoneFileCache *ZoneFileCache) prepare() {
 	zoneFileCache.initCustomFields()
 	zoneFileCache.generateHeaderPrefix()
 	zoneFileCache.generateHeaderSuffix()
-	zoneFileCache.soaSerial = 0
+	zoneFileCache.soaSerial = initSoaSerial(zoneFileNamePrefix + zoneFileCache.domain)
 	zoneFileCache.header = zoneFileCache.generateHeader()
-	zoneFileCache.content = zoneFileCache.header
```
[the remark this answers is not in the document]
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go:58]: I just noticed that it's unnecessary. Not related to this PR.

File: pkg/zonemgr/zone_file.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:21]: You can strive to make it as an API or so please (as Alona suggested https://github.com/kubevirt/kubesecondarydns/pull/45#pullrequestreview-1212967459)

**Comment by dteplits** [comment]
> > Regarding the unit tests, I will change it to use public methods only.
> 
> Unit test can check private methods as well

Then I don't see a problem :)
The only private method I use outside its file is ZoneFileCache.initSoaSerial() -  in unit test
Anyway, I can use a public one instead if necessary.

**Comment by oshoval** [comment]
> > > Regarding the unit tests, I will change it to use public methods only.
> > 
> > 
> > Unit test can check private methods as well
> 
> Then I don't see a problem :) The only private method I use outside its file is ZoneFileCache.initSoaSerial() - in unit test Anyway, I can use a public one instead if necessary.

The problem that Alona raise is the decoupling between the packages IIUC
unrelated to my specific comment about unit tests

**Comment by dteplits** [comment]
> > > > Regarding the unit tests, I will change it to use public methods only.
> > > 
> > > 
> > > Unit test can check private methods as well
> > 
> > 
> > Then I don't see a problem :) The only private method I use outside its file is ZoneFileCache.initSoaSerial() - in unit test Anyway, I can use a public one instead if necessary.
> 
> The problem that Alona raise is the decoupling between the components IIUC unrelated to my specific comment about unit tests. Connection between components is done by the APIs.

So based on the list I specified earlier (public API methods), converting those methods to public should do the job?
This is an API as I define it. Do I miss something?

**Comment by AlonaKaplan** [comment]
> > > > > Regarding the unit tests, I will change it to use public methods only.
> > > > 
> > > > 
> > > > Unit test can check private methods as well
> > > 
> > > 
> > > Then I don't see a problem :) The only private method I use outside its file is ZoneFileCache.initSoaSerial() - in unit test Anyway, I can use a public one instead if necessary.
> > 
> > 
> > The problem that Alona raise is the decoupling between the components IIUC unrelated to my specific comment about unit tests. Connection between components is done by the APIs.
> 
> So based on the list I specified earlier (public API methods), converting those methods to public should do the job? This is an API as I define it. Do I miss something?

Moving `zone_file.go` and `zone_file_cache.go` to package `internal` should block accessing their public methods outside of the package - https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface

**Comment by oshoval** [comment]
> Moving `zone_file.go` and `zone_file_cache.go` to package `internal` should block accessing their public methods outside of the package - https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface

Imo we can just use Cap and non Cap
draw 3 boxes, one for each module, and just the functions that have cross talk declare as API and expose.
The internal can be the next step if needed, and then it might be easier because each module will also have its own folder
(as the internal is based on folder structure)

File: pkg/zonemgr/zone_file_cache_test.go (hunk 4 above)
Comment (AlonaKaplan) [reply pkg/zonemgr/zone_file_cache_test.go:54]: Doesn't seem this comment was handled.

File: pkg/zonemgr/zone_file.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/zonemgr/zone_file.go:21]: I prefer the previous design. I find it cleaner that the ZoneCache is not familiar with the ZoneFile.

File: pkg/zonemgr/zone_file_cache.go [hunk 8]
```diff
@@ -53,12 +56,33 @@ func (zoneFileCache *ZoneFileCache) prepare() {
...
+	if result := soaSerialReg.FindStringSubmatch(content); result != nil && len(result) > 0 {
+		soaSerial := result[1]
+		if soaSerialInt, err := strconv.Atoi(soaSerial); err == nil {
+			return soaSerialInt
+		} else {
+			return 0
+		}
+	} else {
+		return 0
+	}
```
Comment (oshoval) [inline pkg/zonemgr/zone_file_cache.go:83]: Lets try please to reduce else and ident
```
serial := 0
if result := soaSerialReg.FindStringSubmatch(content); result != nil && len(result) > 0 {
    soaSerial := result[1]
    if soaSerialInt, err := strconv.Atoi(soaSerial); err == nil {
        serial = soaSerialInt
    } 
}
return serial
```
same can be done for the function above please

File: pkg/zonemgr/zone_file.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:21]: ZoneFile is a general util it is totally fine imo, 
there should not be two places that determine the serial and rebuild the header
the flow was much harder to grasp

**Comment by dteplits** [comment]
> > > > > > Regarding the unit tests, I will change it to use public methods only.
> > > > > 
> > > > > 
> > > > > Unit test can check private methods as well
> > > > 
> > > > 
> > > > Then I don't see a problem :) The only private method I use outside its file is ZoneFileCache.initSoaSerial() - in unit test Anyway, I can use a public one instead if necessary.
> > > 
> > > 
> > > The problem that Alona raise is the decoupling between the components IIUC unrelated to my specific comment about unit tests. Connection between components is done by the APIs.
> > 
> > 
> > So based on the list I specified earlier (public API methods), converting those methods to public should do the job? This is an API as I define it. Do I miss something?
> 
> Moving `zone_file.go` and `zone_file_cache.go` to package `internal` should block accessing their public methods outside of the package - https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface

will be fixed in a separate PR

File: pkg/zonemgr/zone_file.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/zonemgr/zone_file.go:21]: Why not reading the SOA from the file via ZoneManager and passing it to the ZoneCache?

[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:21]: This is an option, this is not the previous design 
as long as we set initial serial once and build the header once without rebuild in case a serial exists, it is good imo

[the remark this answers is not in the document]
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:21]: The problem is with the zone file name. It is initiated based on the domain name. Which is set inside the ZoneCache. So in current design you can't read the SOA from the file and pass it to ZoneCache without first initializing the ZoneCache. Perhaps it can be changed by changing the design slightly, not sure it's worth it

File: pkg/zonemgr/zone_file_cache_test.go (hunk 4 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache_test.go:55]: fixed

File: pkg/zonemgr/zone_file_cache.go (hunk 8 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go]: fixed

File: pkg/zonemgr/zone_file.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:21]: You can read the SOA from the file because you only look for `SOA.* ( ....` right ?
the regex doesn't depend on the ORIGIN / domain name
if so, the new helper function can be an API of ZoneFile and used by ZoneManager
unless i miss something

[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:21]: You can find the only file in the folder, if exist by scanning the folder
if we all like the idea
then you don't need the name of it
if you want to bullet proof it, you can find the only file that starts with db in the folder etc

File: pkg/zonemgr/zone_file_cache.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/zonemgr/zone_file_cache.go:58]: Now it's back?

File: pkg/zonemgr/zone_manager.go [hunk 9]
```diff
@@ -34,12 +40,37 @@ func NewZoneManager() *ZoneManager {
...
+	soaSerial := zoneMgr.initSoaSerial()
+
 	nameServerIP := os.Getenv(envVarNameServerIP)
 
-	zoneMgr.zoneFileCache = NewZoneFileCache(nameServerIP, domain)
+	zoneMgr.zoneFileCache = NewZoneFileCache(nameServerIP, domain, soaSerial)
+}
 
-	zoneMgr.zoneFile = NewZoneFile(zoneFileName + zoneMgr.zoneFileCache.domain)
+func (zoneMgr *ZoneManager) initSoaSerial() int {
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_manager.go:57]: I think this method should be moved to `ZoneFile`. I would call it `ReadSoaSerial`. If the file doesn't exist it should return `nil`, otherwise, the value from the file.

File: pkg/zonemgr/zone_manager.go [hunk 10]
```diff
@@ -34,12 +40,37 @@ func NewZoneManager() *ZoneManager {
...
+	serial := 0
+	if content, err := zoneMgr.zoneFile.readFile(); err == nil && content != nil {
+		serial = fetchSoaSerial(string(content))
+	}
+	return serial
+}
+
+func fetchSoaSerial(content string) int {
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_manager.go:65]: same

File: pkg/zonemgr/zone_file_test.go [hunk 11]
```diff
@@ -42,15 +43,18 @@ var _ = Describe("disk zone file maintenance", func() {
...
 		AfterEach(func() {
 			Expect(os.RemoveAll(zoneFileName)).To(Succeed())
 		})
 
 		It("should override a zone file on disk", func() {
+			Expect(os.WriteFile(zoneFileName, []byte(zoneFileContent), zoneFilePerm)).To(Succeed())
 			testFileFunc(zoneFileUpdatedContent)
 		})
+
+		It("should read SOA serial", func() {
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_file_test.go:55]: Please add a test testing `ReadSoaSerial` to the scenario the file doesn't exist.

File: pkg/zonemgr/zone_file.go [hunk 12]
```diff
@@ -19,3 +23,26 @@ func NewZoneFile(fileName string) *ZoneFile {
 func (zoneFile *ZoneFile) writeFile(content string) (err error) {
 	return os.WriteFile(zoneFile.zoneFileFullName, []byte(content), zoneFilePerm)
 }
+
+func (zoneFile *ZoneFile) readFile() ([]byte, error) {
+	return os.ReadFile(zoneFile.zoneFileFullName)
+}
+
+func (zoneFile *ZoneFile) ReadSoaSerial() int {
+	serial := 0
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_file.go:32]: I would return `nil` in case the file doesn't exist. The logic to set 0 in cases there is no value in the existing file should reside in the ZoneCache.

File: pkg/zonemgr/zone_file.go [hunk 13]
```diff
@@ -19,3 +23,26 @@ func NewZoneFile(fileName string) *ZoneFile {
...
+	if content, err := zoneFile.readFile(); err == nil && content != nil {
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_file.go:33]: We shouldn't ignore errors. If the error is that the file doesn't exist, it is ok, and we should return `nil` soaSerial. But for any other error we should return an error and not proceed as if nothing happened.

File: pkg/zonemgr/zone_file_cache_test.go [hunk 14]
```diff
@@ -160,7 +159,7 @@ var _ = Describe("cached zone file content maintenance", func() {
 
 		When("interfaces records list contains multiple vmis", func() {
 			BeforeEach(func() {
-				zoneFileCache = NewZoneFileCache(nameServerIP, domain)
+				zoneFileCache = NewZoneFileCache(nameServerIP, domain, 0)
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_file_cache_test.go:162]: Please add a test with non zero value.

File: pkg/zonemgr/zone_file.go (hunk 12 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:32]: I can't return nil for the type int or string but I can return empty string if the file does not exist. Is that ok with you?

File: pkg/zonemgr/zone_file_cache_test.go (hunk 14 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache_test.go]: done

File: pkg/zonemgr/zone_file.go (hunk 13 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go]: done

File: pkg/zonemgr/zone_file.go (hunk 12 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go]: done

File: pkg/zonemgr/zone_file_test.go (hunk 11 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_test.go:61]: done

File: pkg/zonemgr/zone_manager.go (hunk 10 above)
Comment (dteplits) [reply pkg/zonemgr/zone_manager.go]: done

File: pkg/zonemgr/zone_manager.go (hunk 9 above)
Comment (dteplits) [reply pkg/zonemgr/zone_manager.go]: done

File: pkg/zonemgr/zone_file.go [hunk 15]
```diff
@@ -19,3 +25,42 @@ func NewZoneFile(fileName string) *ZoneFile {
...
+
+func (zoneFile *ZoneFile) isFileExist() (*bool, error) {
+	var err error
+	isExist := false
+	if _, err = os.Stat(zoneFile.zoneFileFullName); err == nil {
+		isExist = true
+	} else if errors.Is(err, os.ErrNotExist) {
+		err = nil
+	}
+	return &isExist, err
```
Comment (oshoval) [inline pkg/zonemgr/zone_file.go:41]: why we need *bool and not just bool ?

File: pkg/zonemgr/zone_file.go [hunk 16]
```diff
@@ -19,3 +25,42 @@ func NewZoneFile(fileName string) *ZoneFile {
...
+}
+
+func (zoneFile *ZoneFile) ReadSoaSerial() (*int, error) {
+	if isFileExist, err := zoneFile.isFileExist(); !*isFileExist || err != nil {
+		return nil, err
+	}
+	if content, err := zoneFile.readFile(); content == nil || err != nil {
```
Comment (oshoval) [inline pkg/zonemgr/zone_file.go:48]: worth to add please the logic of isFileExist to readFile 
wdyt ?

File: pkg/zonemgr/zone_file.go [hunk 17]
```diff
@@ -19,3 +25,42 @@ func NewZoneFile(fileName string) *ZoneFile {
...
+		return nil, err
+	} else {
+		return fetchSoaSerial(string(content))
+	}
+}
+
+func fetchSoaSerial(content string) (*int, error) {
```
Comment (oshoval) [inline pkg/zonemgr/zone_file.go:55]: why we need *int not just int ?

please use uint32

soaSerialInt can be named soaSerial

File: pkg/zonemgr/zone_file_cache_test.go [hunk 18]
```diff
@@ -111,7 +117,7 @@ var _ = Describe("cached zone file content maintenance", func() {
 
 		When("interfaces records list contains single vmi", func() {
 			BeforeEach(func() {
-				zoneFileCache = NewZoneFileCache(nameServerIP, domain)
+				zoneFileCache = NewZoneFileCache(nameServerIP, domain, nil)
```
Comment (AlonaKaplan) [inline pkg/zonemgr/zone_file_cache_test.go:120]: I would change at least one on the next tests to use non nil soa number.

File: pkg/zonemgr/zone_file_cache.go [hunk 19]
```diff
@@ -40,10 +39,16 @@ type ZoneFileCache struct {
 	vmiRecordsMap map[string][]string
 }
 
-func NewZoneFileCache(nameServerIP string, domain string) *ZoneFileCache {
+func NewZoneFileCache(nameServerIP string, domain string, soaSerial *int) *ZoneFileCache {
```
Comment (oshoval) [inline pkg/zonemgr/zone_file_cache.go:42]: why *int and not just int ?

File: pkg/zonemgr/zone_file_cache.go [hunk 20]
```diff
@@ -40,10 +39,16 @@ type ZoneFileCache struct {
...
+	soaSerialInt := 0
```
Comment (oshoval) [inline pkg/zonemgr/zone_file_cache.go:43]: soaSerialInt better to just use the argument soaSerial as is, and it will have (does it already ? if not please check) the value zero if the fetch didnt find a previous serial

File: pkg/zonemgr/zone_manager.go [hunk 21]
```diff
@@ -27,19 +28,27 @@ type ZoneManager struct {
...
+	}
+	zoneFileName := zoneFileNamePrefix + domain
+	zoneMgr.zoneFile = NewZoneFile(zoneFileName)
 
-	zoneMgr.zoneFileCache = NewZoneFileCache(nameServerIP, domain)
-
-	zoneMgr.zoneFile = NewZoneFile(zoneFileName + zoneMgr.zoneFileCache.domain)
+	if soaSerial, err := zoneMgr.zoneFile.ReadSoaSerial(); err != nil {
+		return err
+	} else {
```
Comment (oshoval) [inline pkg/zonemgr/zone_manager.go:48]: the else can be dropped, please simply do the code in it, because the if caught it had a return

File: pkg/zonemgr/zone_manager_test.go [hunk 22]
```diff
@@ -21,7 +21,9 @@ var _ = Describe("Zone Manager functionality", func() {
 	BeforeEach(func() {
 		os.Setenv("DOMAIN", customDomain)
 		os.Setenv("NAME_SERVER_IP", customNSIP)
-		zoneMgr = NewZoneManager()
+		var err error
+		zoneMgr, err = NewZoneManager()
+		Expect(err).To(BeNil())
```
Comment (oshoval) [inline pkg/zonemgr/zone_manager_test.go:26]: ToNot(HaveOccurred())

File: pkg/zonemgr/zone_file_test.go [hunk 23]
```diff
@@ -39,18 +40,29 @@ var _ = Describe("disk zone file maintenance", func() {
...
+			Expect(os.WriteFile(zoneFileName, []byte(headerSoaSerial), 0777)).To(Succeed())
+			soaSerial, err := zoneFile.ReadSoaSerial()
+			Expect(err).To(BeNil())
```
Comment (oshoval) [inline pkg/zonemgr/zone_file_test.go:64]: ToNot(HaveOccurred())

**Review by oshoval — COMMENTED** [review]
Thanks

File: pkg/zonemgr/zone_file.go (hunk 16 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:48]: I considered it and preferred to separate them as 1. each function has a different responsibility 2. more readable and clear

File: pkg/zonemgr/zone_file.go (hunk 17 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:55]: since I have to be able to return nil

File: pkg/zonemgr/zone_file_cache.go (hunk 19 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go:42]: since I want to support nil value

File: pkg/zonemgr/zone_file_cache.go (hunk 20 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go:43]: the idea is to handle nil value if ZoneFile not exist

File: pkg/zonemgr/zone_file.go (hunk 16 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:48]: I understand, but on the other hand one can argue readFile should handle non existing file

File: pkg/zonemgr/zone_file.go (hunk 17 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:55]: just return 0, why we need nil ?

File: pkg/zonemgr/zone_file_cache.go (hunk 19 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file_cache.go:42]: just return 0, why we need nil ?
(same for all places)

File: pkg/zonemgr/zone_file_cache.go (hunk 20 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file_cache.go:43]: but 0 is good enough, as long as there is no error, we don't need to differentiate between nil and 0

File: pkg/zonemgr/zone_file.go (hunk 16 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:48]: in this case it returns content=nil, just like os.ReadFile()

File: pkg/zonemgr/zone_file_cache.go (hunk 20 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go:43]: https://github.com/kubevirt/kubesecondarydns/pull/45#discussion_r1054124649

File: pkg/zonemgr/zone_file.go (hunk 17 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:55]: https://github.com/kubevirt/kubesecondarydns/pull/45#discussion_r1054124649

File: pkg/zonemgr/zone_file.go (hunk 16 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file.go:48]: if `readFile` would check before that `isFileExist` it will know to return content == nil if file doesn't exists, with err == nil, and different error in case it couldn't read the file contents

File: pkg/zonemgr/zone_file.go (hunk 12 above)
Comment (oshoval) [reply pkg/zonemgr/zone_file.go]: Imo it is not needed, we don't need to differentiate and have logic for both nil and 0 
just 0 is enough

File: pkg/zonemgr/zone_file_cache_test.go (hunk 18 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache_test.go:120]: I'll add a new test for it

File: pkg/zonemgr/zone_file_cache.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dteplits) [reply pkg/zonemgr/zone_file_cache.go:58]: removed

File: pkg/zonemgr/zone_manager.go [hunk 24]
```diff
@@ -27,19 +28,27 @@ type ZoneManager struct {
...
+	if customDomain := os.Getenv(envVarDomain); customDomain != "" {
+		domain = fmt.Sprintf("%s.%s", domain, customDomain)
+	}
+	zoneFileName := zoneFileNamePrefix + domain
+	zoneMgr.zoneFile = NewZoneFile(zoneFileName)
 
-	zoneMgr.zoneFileCache = NewZoneFileCache(nameServerIP, domain)
-
-	zoneMgr.zoneFile = NewZoneFile(zoneFileName + zoneMgr.zoneFileCache.domain)
+	if soaSerial, err := zoneMgr.zoneFile.ReadSoaSerial(); err != nil {
```
[the remark this answers is not in the document]
Comment (dteplits) [reply pkg/zonemgr/zone_manager.go]: fixed

File: pkg/zonemgr/zone_file.go (hunk 15 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go]: fixed

File: pkg/zonemgr/zone_manager.go (hunk 21 above)
Comment (dteplits) [reply pkg/zonemgr/zone_manager.go]: changed following https://github.com/kubevirt/kubesecondarydns/pull/45#discussion_r1054306875

File: pkg/zonemgr/zone_manager_test.go (hunk 22 above)
Comment (dteplits) [reply pkg/zonemgr/zone_manager_test.go]: done

File: pkg/zonemgr/zone_file_test.go (hunk 23 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file_test.go]: done

File: pkg/zonemgr/zone_file.go (hunk 12 above)
Comment (AlonaKaplan) [reply pkg/zonemgr/zone_file.go]: I don't agree. `ReadSoaSerial` should only read the value from the file, it shouldn't have any heuristic  of calculating the value in case the file doesn't exist, this logic should reside in the `ZoneCache` like the rest of the logic.

Comment (oshoval) [reply pkg/zonemgr/zone_file.go]: you can name the function with something like CalculateSoaSerial
and then it will be responsible to give either 0 or nil, 
otherwise we are drifting with conversions and ptr that we can and should eliminate

i wont insist, but it is not correct imo, keep it simple is much better
maybe there is a middle way you can think of that will make it nicer and you will like ?

File: pkg/zonemgr/zone_file.go (hunk 16 above)
Comment (dteplits) [reply pkg/zonemgr/zone_file.go:48]: I would leave it as it is if you don't mind, I find it more readable

**Review by oshoval — APPROVED** [review]
Thanks

File: pkg/zonemgr/zone_file_test.go [hunk 25]
```diff
@@ -39,18 +40,29 @@ var _ = Describe("disk zone file maintenance", func() {
 		It("should create a zone file on disk", func() {
 			testFileFunc(zoneFileContent)
 		})
+
+		It("should return nil when try to read SOA serial", func() {
+			soaSerial, err := zoneFile.ReadSoaSerial()
+			Expect(err).To(BeNil())
```
Comment (oshoval) [inline pkg/zonemgr/zone_file_test.go:46]: Please consider fixing this one on either this PR or a follow one
Expect(err).ToNot(HaveOccurred())

Comment (dteplits) [reply pkg/zonemgr/zone_file_test.go]: fixed

**Review by oshoval — APPROVED** [review]

**Comment by oshoval** [comment]
I think we didn't cover the following case:
Status-monitor container crashed, for example if someone removed kubevirt for more than 2 minutes
and then deleted all the VMs
Now reinstall kubevirt, and the container is starting with the previous SOA, but it doesnt write the file
as long as there is no update, and the zone file still have the old VMs
the safest solution is to rewrite the file without vms if the soa was there
